design/backup_cancellation.md

```diff
@@ -0,0 +1,154 @@
+
+# Backup Cancellation Design
+
+## Abstract
+This proposal introduces user-initiated backup cancellation functionality to Velero, allowing users to abort running backups through a new `cancel` field in the backup specification.
+The design addresses GitHub issues [#9189](https://github.com/vmware-tanzu/velero/issues/9189
+) and [#2098](https://github.com/vmware-tanzu/velero/issues/2098) by providing a mechanism to cleanly cancel async operations and prevent resource leaks when backups need to be terminated.
+
+## Background
+Currently, Velero lacks the ability to cancel running backups, leading to several critical issues.
+When users accidentally submit broad backup jobs (e.g., forgot to narrow resource selectors), the system becomes blocked and scheduled jobs accumulate.
+Additionally, the backup deletion controller prevents running backups from being deleted.
+
+
+## Goals
+- Enable users to cancel running backups through a `cancel` field in the backup specification
+- Cleanly cancel all associated async operations (BackupItemAction operations, DataUploads, PodVolumeBackups)
+- Provide clear backup phase transitions (InProgress → Cancelling → Cancelled)
+
+## Non Goals
+- Cancelling backups that have already completed or failed
+- Rolling back partially completed backup operations
+- Implementing cancellation for restore operations (future work)
+
+
+## High-Level Design
+The solution introduces a new `cancel` boolean field to the backup specification that users can set to `true` to request cancellation.
+Existing controllers (backup_controller, backup_operations_controller, backup_finalizer_controller) will check for this field and transition the backup to a `Cancelling` phase before returning early from their reconcile loops.
+
+A new dedicated backup cancellation controller will watch for backups in the `Cancelling` phase and coordinate the actual cancellation work.
+This controller will call `Cancel()` methods on all in-progress BackupItemAction operations (which automatically handles DataUpload cancellation), directly cancel PodVolumeBackups by setting their cancel flags, and finally transition the backup to `Cancelled` phase.
```

Comment:
This is specifically related to the asynchronous phase of BIAs, since the Execute() calls will no longer be in progress once the backup reconciler exits. I would replace "on all in-progress BackupItemOperations" with "on all incomplete asynchronous BackupItemAction operations".

Reply:
Also, since PodVolumeBackups are handled synchronously, I don't think any of these will still be InProgress by the time the backup gets to the Cancelling" phase. Within the item backupper, upon completing an ItemBlock, we call `waitUntilPVBsProcessed`, checking PVB status every 5 seconds. You probably need a cancellation check here and if cancelled, set cancelled on PVB and exit the ItemBlock backup. So PVB cancellation will happen in the backup reconciler rather than in the cancellation reconciler.

Reply:
Within `BackupWithResolvers`, when we detect cancellation we break out of the Item -> ItemBlock processing loop
I then plan to get and set the cancel spec on PodVolumeBackups associated with the backup within BackupWIthResolvers when cancellation is detected
`waitUntilPVBsProcessed` checks for either completion or cancellation phase